main/modes/paint/paint_draw.c

```diff
@@ -1120,44 +1139,73 @@ void paintDrawScreenTouchCb(const touch_event_t* evt)
-        {
-            PAINT_LOGD("Swipe RIGHT (%d)", (paintState->lastTouch - paintState->firstTouch) / 32);
-            paintDecBrushWidth((paintState->lastTouch - paintState->firstTouch) / 32);
-        }
-        else if (paintState->firstTouch > paintState->lastTouch)
-        {
-            PAINT_LOGD("Swipe LEFT (%d)", (paintState->firstTouch - paintState->lastTouch + 1) / 32);
-            paintIncBrushWidth((paintState->firstTouch - paintState->lastTouch + 1) / 32);
-        }
-        else
+        case BTN_MODE_DRAW:
+        case BTN_MODE_SELECT:
         {
-            // TAP
-            PAINT_LOGD("Tap pad %d", evt->pad);
-            if (evt->pad == 0)
+            if (evt->down && evt->state != 0 && !paintState->touchDown)
             {
-                // Tap X (?)
-                paintIncBrushWidth(1);
+                paintState->touchDown = true;
+                paintState->firstTouch = evt->position;
+                paintState->lastTouch = evt->position;
+
+                paintEnterSelectMode();
+            }
+            else if (evt->down && evt->state != 0 && paintState->touchDown) {
+                paintState->lastTouch = evt->position;
             }
-            else if (evt->pad == 4)
+            else if (!evt->down && evt->state == 0 && paintState->touchDown)
             {
-                // Tap Y (?)
-                paintDecBrushWidth(1);
+                if (paintState->firstTouch + 32 < paintState->lastTouch)
+                {
+                    PAINT_LOGD("Swipe RIGHT (%d)", (paintState->lastTouch - paintState->firstTouch) / 32);
+                    paintDecBrushWidth((paintState->lastTouch - paintState->firstTouch) / 32);
+                }
+                else if (paintState->firstTouch > paintState->lastTouch + 32)
+                {
+                    PAINT_LOGD("Swipe LEFT (%d)", (paintState->firstTouch - paintState->lastTouch + 1) / 32);
+                    paintIncBrushWidth((paintState->firstTouch - paintState->lastTouch + 1) / 32);
+                }
+                else if (paintState->firstTouch == paintState->lastTouch)
+                {
+                    // TAP
+                    if (evt->pad == 0)
+                    {
+                        // Tap X (?)
+                        PAINT_LOGD("Tap X");
+                        paintIncBrushWidth(1);
+                    }
+                    else if (evt->pad == 4)
+                    {
+                        // Tap Y (?)
+                        PAINT_LOGD("Tap Y");
+                        paintDecBrushWidth(1);
+                    }
+                }
+
+                paintExitSelectMode();
+                paintState->touchDown = false;
             }
+
+            break;
         }
-        paintState->touchDown = false;
+
+        case BTN_MODE_PALETTE:
+        {
+            int32_t centroid, intensity;
+            if (getTouchCentroid(&centroid, &intensity))
```

Comment:
This doesn't work the way you think it works. The callback is only called when a pad is touched or released, but for maximum smoothness, this should be called somewhere else and more frequently (`paintMainLoop()` perhaps). I want to see the brush modifies (x1, x2, x3) update in realtime as I swipe!

Reply:
heh, I thought maybe I could get away with it like this but yeah it does make it very choppy...

Reply:
Now it's suuuuper smooth! And the brush size updates in real-time too

Reply:
Much better!